src/main/scala/tech/cryptonomic/conseil/generic/chain/MultiFetchDecoding.scala

```diff
@@ -0,0 +1,209 @@
+package tech.cryptonomic.conseil.generic.chain
+
+import cats._
+import cats.data.Kleisli
+import cats.syntax.semigroupal._
+
+/** Contract for a generic combination of operations that get encoded data (usually json) from
+  * some resource (e.g. the tezos node), then converts that to a value, batching multiple requests together.
+  * The returned values are tupled with the corresponding input
+  * The interface assumes some kind of higher-order typed effect `F[_]` is part
+  * of the operation execution, and represents everything parametrically, to leave
+  * room for future evolutions of the code protocol.
+  * Thus the required implementation is in the form of a `cats.data.Kleisli` type,
+  * parametric on effect `F`, input `In`, output `Out`. As needed, the in/out types are
+  * wrapped in a `T[_]` type constructor that represents multiple values (e.g. a collection type).
+  *
+  * @tparam F the higher-order effect type (e.g. `Try`, `Future`, `IO`, `Either`)
+  * @tparam T a traversable type (used to wrap mutliple requests and corresponding return values)
+  */
+trait MultiFetchDecoding[F[_], T[_]] {
```

Comment:
For my education, could you please describe why you chose to create a typeclass instead of a regular trait? 

When I look at this in the most minimal manner possible, I can certainly see the point of the following:
- Defining a trait with common workflow functions will unite all blockchain implementations.
- Using Kleisli to make the composition of the workflow functions cleaner to read.

However, beyond this, I am not able to see the additional advantage of defining an entire typeclass. Generally, as I understand it, typeclasses allow one to change the behaviour of categories of types without having to change their implementations, e.g. defining `toJSON()` functionality for multiple types like strings, integers, monkeys, bananas, etc. In this case, however, we are not enhancing the behaviour of any existing types but simply defining a few new types for various workflows. If the following were done using typeclasses, it would make intuitive sense to me:

```scala
val levels = Set(773, 774, 775, 776, 777)
val blockData = levels.fetchTezosBlocks()
```

Could you please explain a bit further the additional utility of implementing a typeclass in this specific way? :smile:

Reply:
In this case we're only using an interface... to be a type class we should have defined implicit instances for the custom types, to use as constraints in other method calls.

In fact here the usage pattern is that of a traditioal OO interface. This is standard information hiding used to provide common higher-level interaction with tezos nodes batched interaction and json parsing, with some reuse.

It allows paves the road to a potential switch to other async models that we can explore soon.
Examples might be 
- using Akka-Streams by changing Future[List[Block]] to Source[Id[Block]] (which is actually `Source[Block]`)
- using monix/cats IO changing Future[List[Block]] to ??? I don't know exactly yet
- using zio using ZIO.Task[ZIO.Stream[Block]] or something alike

Reply:
So you're saying the advantage here, if I'm understanding correctly, is to generalize over the effects that various workflows over blockchain endpoints may require?